crates/nu-cli/src/util.rs

```diff
@@ -199,17 +204,126 @@ fn gather_env_vars(
     }
 }
 
-pub fn eval_source(
+pub fn make_main_call(
+    engine_state: &mut EngineState,
+    fname: String,
+    args: &[String],
+    redirect_env: bool,
+) -> Result<Arc<Block>, ShellError> {
+    let mut working_set = StateWorkingSet::new(engine_state);
+    let source = format!("{} {}", fname, args.join(" "));
+    let source = source.as_bytes();
+    let file_id = working_set.add_file("<commandline>".to_string(), source);
+    let file_span = working_set.get_span_for_file(file_id);
+    let (tokens, err) = lex(source, file_span.start, &[], &[], false);
+    if let Some(err) = err {
+        working_set.error(err)
+    }
+    let (lite_block, err) = lite_parse(tokens.as_slice());
+    if let Some(err) = err {
+        working_set.error(err);
+    }
+    let lite_command = &lite_block.block[0].commands[0];
+
+    let (decl_id, command_len) =
+        find_longest_command(&working_set, b"main", &lite_command.parts[1..])
+            .expect("already checked that a main def exists");
```

Comment:
I think this is fine as a panic, it's a developer error, but I would suggest something like `"make_main_call() called, but 'main' definition not found in state"`